libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -881,6 +915,75 @@ string YulUtilFunctions::readFromStorageDynamic(Type const& _type, bool _splitFu
 	});
 }
 
+string YulUtilFunctions::readFromMemory(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, false);
+}
+
+string YulUtilFunctions::readFromCalldata(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, true);
+}
+
+string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromCalldata)
+{
+	string functionName =
+		string("read_from_") +
+		(_fromCalldata ? "calldata" : "memory") +
+		"_dynamic_" +
+		_type.identifier();
+
+	return m_functionCollector->createFunction(functionName, [&] {
+		if (auto arrayType = dynamic_cast<ArrayType const*>(&_type))
```

Comment:
If `_type.isValueType()` is false, then it should always be just `mload` (if from memory) without any pre- or post-processing. You can add assertions that the size on stack is equal to 1.

Reply:
instead of the array condition/conversion here just check for isValueType() then?

Reply:
Perhaps check for ReferenceType and then assert that in the other case, it is a value type.